src/core/timer.ml

```diff
@@ -28,45 +29,57 @@ let get_time = Extc.time
 let htimers = Hashtbl.create 0
 
 let new_timer id =
+	let now = get_time() in
 	let key = String.concat "." id in
 	try
 		let t = Hashtbl.find htimers key in
-		t.start <- get_time() :: t.start;
+		t.start <- now :: t.start;
+		t.pauses <- 0. :: t.pauses;
 		t.calls <- t.calls + 1;
 		t
 	with Not_found ->
-		let t = { id = id; start = [get_time()]; total = 0.; calls = 1; } in
+		let t = { id = id; start = [now]; pauses = [0.]; total = 0.; calls = 1; } in
 		Hashtbl.add htimers key t;
 		t
 
 let curtime = ref []
 
-let close t =
-	let start = (match t.start with
-		| [] -> assert false
-		| s :: l -> t.start <- l; s
-	) in
-	let now = get_time() in
-	let dt = now -. start in
-	t.total <- t.total +. dt;
-	let rec loop() =
-		match !curtime with
-		| [] -> failwith ("Timer " ^ (String.concat "." t.id) ^ " closed while not active")
-		| tt :: l -> curtime := l; if t != tt then loop()
-	in
-	loop();
+let rec close now t =
+	match !curtime with
+	| [] ->
+		failwith ("Timer " ^ (String.concat "." t.id) ^ " closed while not active")
+	| tt :: rest ->
+		if t == tt then begin
+			let start = List.hd t.start in
```

Comment:
This can be done a bit more elegantly by matching `t.start` against `hd :: tl` because we use the tail further below.